src/Columns/ColumnSparse.h

```diff
@@ -0,0 +1,223 @@
+    bool isCollationSupported() const override { return values->isCollationSupported(); }
+
+    size_t getNumberOfDefaults() const { return _size - offsets->size(); }
+    size_t getNumberOfTrailingDefaults() const
+    {
+        return offsets->empty() ? _size : _size - getOffsetsData().back() - 1;
+    }
+
+    /// Return position of element in 'values' columns,
+    /// that corresponds to n-th element of full column.
+    /// O(log(size)) complexity,
+    size_t getValueIndex(size_t n) const;
+
+    const IColumn & getValuesColumn() const { return *values; }
+    IColumn & getValuesColumn() { return *values; }
+
+    const ColumnPtr & getValuesPtr() const { return values; }
+    ColumnPtr & getValuesPtr() { return values; }
+
+    const IColumn::Offsets & getOffsetsData() const;
+    IColumn::Offsets & getOffsetsData();
+
+    const ColumnPtr & getOffsetsPtr() const { return offsets; }
+    ColumnPtr & getOffsetsPtr() { return offsets; }
+
+    const IColumn & getOffsetsColumn() const { return *offsets; }
+    IColumn & getOffsetsColumn() { return *offsets; }
+
+    /// This class helps to iterate over all values in ColumnSparse.
+    class Iterator
+    {
+    public:
+        Iterator(const PaddedPODArray<UInt64> & offsets_, size_t size_, size_t current_offset_, size_t current_row_)
+            : offsets(offsets_), size(size_), current_offset(current_offset_), current_row(current_row_)
+        {
+        }
+
+        bool ALWAYS_INLINE isDefault() const { return current_offset == offsets.size() || current_row != offsets[current_offset]; }
+        size_t ALWAYS_INLINE getValueIndex() const { return isDefault() ? 0 : current_offset + 1; }
+        size_t ALWAYS_INLINE getCurrentRow() const { return current_row; }
+        size_t ALWAYS_INLINE getCurrentOffset() const { return current_offset; }
+
+        bool operator==(const Iterator & other) const
+        {
+            return size == other.size
+                && current_offset == other.current_offset
+                && current_row == other.current_row;
+        }
+
+        bool operator!=(const Iterator & other) const { return !(*this == other); }
+
+        Iterator operator++()
+        {
+            if (!isDefault())
+                ++current_offset;
+            ++current_row;
+            return *this;
+        }
+
+    private:
+        const PaddedPODArray<UInt64> & offsets;
+        const size_t size;
+        size_t current_offset;
+        size_t current_row;
+    };
+
+    Iterator begin() const { return Iterator(getOffsetsData(), _size, 0, 0); }
+    Iterator end() const { return Iterator(getOffsetsData(), _size, getOffsetsData().size(), _size); }
+
+private:
+    /// Contains default value at 0 position.
+    WrappedPtr values;
+
+    /// Sorted offsets of non-default values in the full column.
+    /// 'offsets[i]' corresponds to 'values[i + 1]'.
```

Comment:
As I understand, it is because we store default at 0 position in `values`.
Or, like, values is shifted by 1.
Why it is reasonable?

Reply:
Ok, I see that it may be convenient to store all possible values in a single column.
Maybe add some comment why

Reply:
I've added a comment.